src/nsrt_learning/side_predicate_learning.py

```diff
@@ -280,3 +280,180 @@ def _get_successor_with_correct_option(
+                                  if p.option_spec[0] == option.parent]
+                assert len(matching_pnads) == 1
+                pnad = matching_pnads[0]
+                # By construction, this segment should exist in this PNAD.
+                var_to_obj = None
+                for pnad_seg, sub in pnad.datastore:
+                    if pnad_seg is segment:
+                        var_to_obj = sub
+                        break
+                assert var_to_obj is not None
+                # Check if the current image is predicted by the pnad. If not,
+                # then we will need to make the pnad effects more specific.
+                current_add_effects = current_image - atoms_seq[t]
+                pnad_ground_add_effects = {a.ground(var_to_obj)
+                                           for a in pnad.op.add_effects}
+                if not current_add_effects.issubset(pnad_ground_add_effects):
+                    # Need to make the effects more specific.
+                    assert len(set(var_to_obj.values())) == len(var_to_obj)
+                    obj_to_var = {o: v for v, o in var_to_obj.items()}
+                    missing_effects = current_add_effects - \
+                        pnad_ground_add_effects
+                    # Unclear if the below is necessary in general, maybe.
+                    # Check if we need to introduce new parameters.
+                    # all_objects = {o for a in missing_effects
+                    #                for o in a.objects}
+                    # unbound_objects = all_objects - set(obj_to_var)
+                    # new_var_idx = 0
+                    # for v in var_to_obj:
+                    #     assert v.name.startswith("?x")
+                    #     new_var_idx = max(new_var_idx, int(v.name[2:])+1)
+                    # for obj in sorted(unbound_objects):
+                    #     new_var = Variable(f"?x{new_var_idx}", obj.type)
+                    #     new_var_idx += 1
+                    #     obj_to_var[obj] = new_var
+                    #     var_to_obj[new_var] = obj
+                    new_add_effects = {a.lift(obj_to_var)
+                                       for a in missing_effects}
+                    param_option_to_spec_and_effects[option.parent]["add"] |= \
+                        new_add_effects
+                    # Update the PNAD based on the new effects.
+                    # TODO: refactor to only update the PNAD that just changed.
+                    # This is horrible.
+                    current_pnads = self._induce_pnads_from_spec_and_effects(
+                        param_option_to_spec_and_effects)
+                    matching_pnads = [p for p in current_pnads
+                                      if p.option_spec[0] == option.parent]
+                    assert len(matching_pnads) == 1
+                    pnad = matching_pnads[0]
+                    pnad_ground_add_effects = {a.ground(var_to_obj)
```

Comment:
wait, why wouldn't this always be `current_add_effects`? you just added the missing (lifted) add effects to the previous pnad ground add effects, so you should end up with `current_add_effects` always, right